cmd/vault.go

```diff
@@ -397,3 +402,326 @@ func readJSON(name string) (secret interface{}, err error) {
+
+	cmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create new user secret",
+		Long:  `Create new user secret`,
+		Example: `
+	privx-cli user-secrets create [access flags] --name <SECRET-NAME>
+		--allow-read-to <ROLE-ID>
+		--allow-write-to <ROLE-ID>
+		...
+		JSON-FILE
+		`,
+		Args:         cobra.ExactArgs(1),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return userSecretCreate(args, options)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.secretName, "name", "", "secret name")
+	flags.StringArrayVar(&options.vaultReadTo, "allow-read-to", []string{}, "read by role ID")
+	flags.StringArrayVar(&options.vaultWriteTo, "allow-write-to", []string{}, "write by role ID")
+	flags.StringVar(&options.ownerID, "owner-id", "", "secret's owner ID")
+	cmd.MarkFlagRequired("owner-id")
+	cmd.MarkFlagRequired("name")
+	cmd.MarkFlagRequired("read-by")
+	cmd.MarkFlagRequired("write-by")
+
+	return cmd
+}
+
+func userSecretCreate(args []string, options vaultOptions) error {
+	secret, err := readJSON(args[0])
+	if err != nil {
+		return err
+	}
+
+	secretID := &vault.SecretID{
+		OwnerID: options.ownerID,
+		Name:    options.secretName,
+	}
+	api := vault.New(curl())
+	if err := api.CreateUserSecret(*secretID, options.vaultReadTo,
+		options.vaultWriteTo, secret); err != nil {
+		return err
+	}
+
+	return stdout(secret)
+}
+
+//
+
+// update user secret
+func userSecretUpdateCmd() *cobra.Command {
+	options := vaultOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "Update user secret",
+		Long:  `Update existing user secret`,
+		Example: `
+	privx-cli user-secrets update [access flags] --owner-id <SECRET-NAME> --name <SECRET-NAME> JSON-FILE
+	privx-cli user-secrets update [access flags] --owner-id <SECRET-NAME> --name <SECRET-NAME>
+		--allow-read-to <ROLE-ID>
+		--allow-write-to <ROLE-ID>
+		...
+		JSON-FILE
+		`,
+		Args:         cobra.ExactArgs(1),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return userSecretUpdate(options, args)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.secretName, "name", "", "secret name")
+	flags.StringArrayVar(&options.vaultReadTo, "allow-read-to", []string{}, "read by role ID")
+	flags.StringArrayVar(&options.vaultWriteTo, "allow-write-to", []string{}, "write by role ID")
+	flags.StringVar(&options.ownerID, "owner-id", "", "secret's owner ID")
+	cmd.MarkFlagRequired("owner-id")
+	cmd.MarkFlagRequired("name")
+
+	return cmd
+}
+
+func userSecretUpdate(options vaultOptions, args []string) error {
+	secret, err := readJSON(args[0])
+	if err != nil {
+		return err
+	}
+
+	secretID := &vault.SecretID{
+		OwnerID: options.ownerID,
+		Name:    options.secretName,
+	}
+	api := vault.New(curl())
+	bag, err := api.UserSecret(*secretID)
+	if err != nil {
+		return err
+	}
+
+	if len(options.vaultReadTo) == 0 {
+		for _, ref := range bag.AllowRead {
+			options.vaultReadTo = append(options.vaultReadTo, ref.ID)
```

Comment:
Usually it is better to declare new variables (to hold processed user input) and leave the user input intact. Do not modify options values. Same suggestion for the options.vaultWriteTo below.